plugins/powerpc/powerpc_branch.ml

```diff
@@ -0,0 +1,277 @@
+open Powerpc.Std
+
+let update_link_register cpu ops =
+  RTL.[cpu.lr := cpu.pc + unsigned const byte 4]
+
+(** Branch Instructions, Branch
+    Page 37 of IBM Power ISATM Version 3.0 B
+    examples:
+    4b ff fe f0  b .+67108592
+    4b ff fe f2  ba 67108592
+    4b ff fe f1  bl .+67108592
+    4b ff fe f3  bla 67108592 *)
+let b cpu ops =
+  let im = signed imm ops.(0) in
+  let tm = signed var (bitwidth 26) in
+  let sh = unsigned of_string "0b00" in
+  RTL.[
+    tm := im ^ sh;
+    cpu.jmp (cpu.pc + tm)
+  ]
+
+let ba cpu ops =
+  let im = signed imm ops.(0) in
+  let tm = signed var (bitwidth 26) in
+  let sh = unsigned of_string "0b00" in
+  RTL.[
+    tm := im ^ sh;
+    cpu.jmp tm;
+  ]
+
+let bl cpu ops =
+  let im = signed imm ops.(0) in
+  let tm = signed var (bitwidth 26) in
+  let sh = unsigned of_string "0b00" in
+  let ad = unsigned const byte 4 in
+  RTL.[
+    tm := im ^ sh;
+    cpu.jmp (cpu.pc + tm);
+    cpu.lr := cpu.pc + ad;
+  ]
+
+let bla cpu ops =
+  let im = signed imm ops.(0) in
+  let tm = signed var (bitwidth 26) in
+  let sh = unsigned of_string "0b00" in
+  let ad = unsigned const byte 4 in
+  RTL.[
+    tm := im ^ sh;
+    cpu.jmp tm;
+    cpu.lr := cpu.pc + ad;
+  ]
+
+(** Branch Instructions, Branch Conditional
+    Page 37 of IBM Power ISATM Version 3.0 B
+    examples:
+    42 9f 00 04  bc 20, 31, .+4
+    42 9f 00 06  bca 20, 31, 4
+    42 9f 00 05  bcl 20, 31, .+4
+    42 9f 00 07  bcla 20, 31, 4   *)
+let bc cpu ops =
+  let bo = unsigned imm ops.(0) in
+  let bi = unsigned cpu.reg ops.(1) in
+  let bd = unsigned imm ops.(2) in
+  let sh = unsigned of_string "0b00" in
+  let ctr_ok = unsigned var bit in
+  let cond_ok = unsigned var bit in
+  let x = unsigned var (bitwidth 5) in
+  let tm = signed var halfword in
+  RTL.[
+    x := last bo 5;
+    when_ (nth bit x 2 = zero) [
+      cpu.ctr := cpu.ctr - one;
+    ];
+    ctr_ok := nth bit x 2 lor ((cpu.ctr <> zero) lxor (nth bit x 3));
+    cond_ok := nth bit x 0 lor (bi lxor (lnot (nth bit x 1)));
+    when_ (ctr_ok land cond_ok) [
+      tm := bd ^ sh;
+      cpu.jmp (cpu.pc + tm);
+    ]
+  ]
+
+let bca cpu ops =
+  let bo = unsigned imm ops.(0) in
+  let bi = unsigned cpu.reg ops.(1) in
+  let bd = unsigned imm ops.(2) in
+  let sh = unsigned of_string "0b00" in
+  let ctr_ok = unsigned var bit in
+  let cond_ok = unsigned var bit in
+  let x = unsigned var (bitwidth 5) in
+  let tm = signed var cpu.word_width in
+  RTL.[
+    x := last bo 5;
+    when_ (nth bit x 2 = zero) [
+      cpu.ctr := cpu.ctr - one;
+    ];
+    ctr_ok := nth bit x 2 lor ((cpu.ctr <> zero) lxor (nth bit x 3));
+    cond_ok := nth bit x 0 lor (bi lxor (lnot (nth bit x 1)));
+    when_ (ctr_ok land cond_ok) [
+      tm := bd ^ sh;
+      cpu.jmp tm;
+    ]
+  ]
+
+let bcl = bc ^ update_link_register
```

Comment:
this won't work, as it will update the link register after the jump, not before.